<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="None">
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="img/favicon.ico">
  <title>Home - PRSice-2</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="css/theme.css" type="text/css" />
  <link rel="stylesheet" href="css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Home";
    var mkdocs_page_input_path = "index.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="js/jquery-2.1.1.min.js" defer></script>
  <script src="js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-113286405-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="." class="icon icon-home"> PRSice-2</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1 current">
		
    <a class="current" href=".">Home</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#executable-downloads">Executable downloads</a></li>
    

    <li class="toctree-l2"><a href="#2020-05-20-v230c">2020-05-20  (v2.3.0.c)</a></li>
    

    <li class="toctree-l2"><a href="#2020-05-19-v230b">2020-05-19 (v2.3.0.b)</a></li>
    

    <li class="toctree-l2"><a href="#2020-05-19-v230a">2020-05-19 (v2.3.0.a)</a></li>
    

    <li class="toctree-l2"><a href="#2020-05-18-v230">2020-05-18 (v2.3.0)</a></li>
    

    <li class="toctree-l2"><a href="#r-packages-requirements">R Packages Requirements</a></li>
    

    <li class="toctree-l2"><a href="#quick-start">Quick Start</a></li>
    

    <li class="toctree-l2"><a href="#output-of-results">Output of Results</a></li>
    

    <li class="toctree-l2"><a href="#detailed-guide">Detailed Guide</a></li>
    

    <li class="toctree-l2"><a href="#full-command-line-options">Full command line options</a></li>
    
        <ul>
        
            <li><a class="toctree-l3" href="#citation">Citation</a></li>
        
            <li><a class="toctree-l3" href="#support">Support</a></li>
        
            <li><a class="toctree-l3" href="#authors">Authors</a></li>
        
            <li><a class="toctree-l3" href="#acknowledgement">Acknowledgement</a></li>
        
        </ul>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Quick Start</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="quick_start/">PRSice</a>
                </li>
                <li class="">
                    
    <a class="" href="quick_start_prset/">PRSet</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Detail Guides</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="step_by_step/">PRSice</a>
                </li>
                <li class="">
                    
    <a class="" href="prset_detail/">PRSet</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="command_detail/">Available Commands</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Developers</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="compilation/">Compile from Source</a>
                </li>
                <li class="">
                    
    <a class="" href="decisions/">Development Decisions</a>
                </li>
                <li class="">
                    
    <a class="" href="resources/">Useful Resources</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Misc</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="extra_steps/">Additional Steps for MAC and Window users</a>
                </li>
                <li class="">
                    
    <a class="" href="faq/">Frequently Asked Questions</a>
                </li>
                <li class="">
                    
    <a class="" href="archive/">Archive</a>
                </li>
                <li class="">
                    
    <a class="" href="update_log/">Update Log</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href=".">PRSice-2</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".">Docs</a> &raquo;</li>
    
      
    
    <li>Home</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRSice/edit/master/docs/index.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1>PRSice-2: Polygenic Risk Score software</h1>

<p>PRSice (pronounced 'precise') is a Polygenic Risk Score software for calculating, applying, evaluating and plotting the results of polygenic risk scores (PRS) analyses. Some of the features include:</p>
<p>PRSice (pronounced 'precise') is a Polygenic Risk Score software for calculating, applying, evaluating and plotting the results of polygenic risk scores (PRS) analyses. Some of the features include:</p>
<ol>
<li>High-resolution scoring (PRS calculated across a large number of P-value thresholds)</li>
<li>Identify Most predictive PRS</li>
<li>Empirical P-values output (not subject to over-fitting)</li>
<li>Genotyped (PLINK binary) and imputed (Oxford bgen v1.2) data input</li>
<li>Biobank-scale genotyped data can be analysed within hours</li>
<li>Incorporation of covariates</li>
<li>Application across multiple target traits simultaneously</li>
<li>Results plotted in several formats (bar plots, high-res plots, quantile plots)</li>
<li>PRSet: function for calculating PRS across user-defined pathways / gene sets</li>
</ol>
<h1 id="executable-downloads">Executable downloads <a href="https://doi.org/10.5281/zenodo.3703335"><img alt="DOI" src="https://zenodo.org/badge/DOI/10.5281/zenodo.3703335.svg" /></a><a href="https://coveralls.io/github/choishingwan/PRSice?branch=master"><img alt="Coverage Status" src="https://coveralls.io/repos/github/choishingwan/PRSice/badge.svg?branch=master" /></a><a class="headerlink" href="#executable-downloads" title="Permanent link">&para;</a></h1>
<table>
<thead>
<tr>
<th>Operating System</th>
<th align="center">Link</th>
</tr>
</thead>
<tbody>
<tr>
<td>Linux 64-bit</td>
<td align="center"><a href="https://github.com/choishingwan/PRSice/releases/download/2.3.0/PRSice_linux.230c.zip">v2.3.0.c</a></td>
</tr>
<tr>
<td>OS X 64-bit</td>
<td align="center"><a href="https://github.com/choishingwan/PRSice/releases/download/2.3.0/PRSice_mac.230c.zip">v2.3.0.c</a></td>
</tr>
<tr>
<td>Windows 32-bit</td>
<td align="center">Not available</td>
</tr>
<tr>
<td>Windows 64-bit</td>
<td align="center">Not available</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Latest Update</p>
<h1 id="2020-05-20-v230c">2020-05-20  (v2.3.0.c)<a class="headerlink" href="#2020-05-20-v230c" title="Permanent link">&para;</a></h1>
<ul>
<li>Fix all score output format</li>
<li>Fix problem with <code>--no-regress</code>. Might still have problem with <code>--no-regress --score con-std</code></li>
</ul>
<h1 id="2020-05-19-v230b">2020-05-19 (v2.3.0.b)<a class="headerlink" href="#2020-05-19-v230b" title="Permanent link">&para;</a></h1>
<ul>
<li>Fix error where sample selection will distort phenotype loading, loading the wrong phenotype to wrong sample. As this is a major bug, we deleted the previous 2 releases. Sorry for the troubles.</li>
</ul>
<h1 id="2020-05-19-v230a">2020-05-19 (v2.3.0.a)<a class="headerlink" href="#2020-05-19-v230a" title="Permanent link">&para;</a></h1>
<ul>
<li>Fix output error where we always say 0 valid phenotype were included for continuous trait</li>
<li>Fix problem with permutation where PRSice will crash if input are rank deficient </li>
<li>Fix problem when provide a binary phenotype file with a fam file containing -9 as phenotype, PRSice will wrongly state that there are no phenotype presented</li>
<li>Fix problem in Rscript where if sample ID is numeric and starts with 0, the best file will not merge with the phenotype file, causing 0 valid PRS to be observed</li>
</ul>
<h1 id="2020-05-18-v230">2020-05-18 (v2.3.0)<a class="headerlink" href="#2020-05-18-v230" title="Permanent link">&para;</a></h1>
<ul>
<li>We now support multi-threaded clumping (separated by chromosome)</li>
<li>Genotypes will be stored to memory during clumping (increase memory usage, significantly speed up clumping)</li>
<li>Will only generate one .prsice file for all phenotypes<ul>
<li>.prsice file now has additional column call "Pheno"</li>
</ul>
</li>
<li>Introduced <code>--chr-id</code> which generate rs id based on user provided formula (see <a href="command_detail/">detail</a> for more info)</li>
<li>Format of <code>--base-maf</code> and <code>--base-info</code> are now changed to <code>&lt;name&gt;:&lt;value&gt;</code> from <code>&lt;name&gt;,&lt;value&gt;</code></li>
<li>Fix a bug related to ambiguous allele dosage flipping when <code>--keep-ambig</code> is used</li>
<li>Better mismatch handling. For example, if your base file only provide the effective allele A without the non-effective allele information, PRSice will now do dosage flipping if your target file has G/C as effective allele and A /T as an non-effective allele (whereas previous this SNP will be considered as a mismatch)</li>
<li>Fix bug in 2.2.13 where PRSice won't output the error message during command parsing stage</li>
<li>If user provided the <code>--stat</code> information, PRSice will now error out instead of trying to look for BETA or OR in the file. </li>
<li>PRSice should now better recognize if phenotype file contains a header</li>
<li>various small bug fix</li>
</ul>
<p>update log for previous release can be found <a href="update_log/">here</a></p>
</div>
<div class="admonition caution">
<p class="admonition-title">Caution</p>
<p>We have now fixed window problem. But was unable to access the computer that is used for compilation due to COVID. Will try to compile it when we regain access.</p>
</div>
<div class="admonition caution">
<p class="admonition-title">Caution</p>
<p>PRSet are currently under open beta - results output are reliable but please report any specific problems to our google group (see Support below)3</p>
</div>
<h1 id="r-packages-requirements">R Packages Requirements<a class="headerlink" href="#r-packages-requirements" title="Permanent link">&para;</a></h1>
<p>To plot graphs, PRSice requires <a href="https://www.r-project.org/">R</a> (<strong>version 3.2.3+</strong>) installed.</p>
<p><a href="extra_steps/">Additional steps</a> might be required for Mac and Windows users.</p>
<div class="admonition note">
<p class="admonition-title">Installing required R packages</p>
<p>PRSice can automatically download all required packages, even without administrative right.
You can specify the install directory using <code>--dir</code>. For example</p>
<div class="codehilite"><pre><span></span>    Rscript PRSice.R --dir .
</pre></div>

<p>will install all required packages under the local directory.</p>
</div>
<h1 id="quick-start">Quick Start<a class="headerlink" href="#quick-start" title="Permanent link">&para;</a></h1>
<p>For Quick start use, please refer to <a href="quick_start/">Quick Start</a></p>
<div class="admonition tip">
<p class="admonition-title">List user options</p>
<p>You can also type</p>
<div class="codehilite"><pre><span></span>    ./PRSice
</pre></div>

<p>to view all available parameters unrelated to plotting, or</p>
<div class="codehilite"><pre><span></span>    Rscript PRSice.R -h
</pre></div>

<p>to view all available parameters, including those used for plotting</p>
</div>
<h1 id="output-of-results">Output of Results<a class="headerlink" href="#output-of-results" title="Permanent link">&para;</a></h1>
<p>You can see the expected output of PRSice <a href="step_by_step/#output-of-results">here</a></p>
<h1 id="detailed-guide">Detailed Guide<a class="headerlink" href="#detailed-guide" title="Permanent link">&para;</a></h1>
<p>You can find a more detailed document explaining the input and output of PRSice in <a href="step_by_step/">this page</a></p>
<h1 id="full-command-line-options">Full command line options<a class="headerlink" href="#full-command-line-options" title="Permanent link">&para;</a></h1>
<p>You can find all command line options of PRSice under the section <em>Details of PRSice/PRSet</em></p>
<h2 id="citation">Citation<a class="headerlink" href="#citation" title="Permanent link">&para;</a></h2>
<p>If you use PRSice, then please cite:</p>
<div class="admonition important">
<p class="admonition-title">Citation</p>
<p>Choi SW, and Oâ€™Reilly PF. "PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data." GigaScience 8, no. 7 (July 1, 2019). <a href="https://doi.org/10.1093/gigascience/giz082">https://doi.org/10.1093/gigascience/giz082</a>.</p>
</div>
<h2 id="support">Support<a class="headerlink" href="#support" title="Permanent link">&para;</a></h2>
<p>This wiki should contain all the basic instruction for the use of PRSice.
Shall you have any problems, please feel free to start an issue <a href="https://github.com/choishingwan/PRSice/issues">here</a> or visit our <a href="https://groups.google.com/forum/#!forum/prsice">google group</a>.
You can help us to speed up the debug process by including the log file generated by PRSice.</p>
<p>In addition, you can use the search bar in this webpage to search for specific functions. </p>
<h2 id="authors">Authors<a class="headerlink" href="#authors" title="Permanent link">&para;</a></h2>
<p>For more details on the authors, see:</p>
<ul>
<li><a href="https:choishingwan.github.io">Dr Shing Wan Choi</a></li>
<li><a href="http://www.pauloreilly.info/">Dr Paul O'Reilly</a></li>
</ul>
<p>PRSice-2 and all new functionalities are coded by:</p>
<ul>
<li><a href="https://choishingwan.github.io">Dr Shing Wan Choi</a></li>
</ul>
<h2 id="acknowledgement">Acknowledgement<a class="headerlink" href="#acknowledgement" title="Permanent link">&para;</a></h2>
<p>PRSice is a software package written in C++ (main) and R (plotting).
The code relies partially on those written in PLINK by <a href="https://www.cog-genomics.org/software">Christopher Chang</a>.
Management of BGEN file is based on BGEN lib written by <a href="https://bitbucket.org/gavinband/bgen">Gavin Band</a>.
We also utilize the <a href="https://eigen.tuxfamily.org">Eigen C++</a> library, the <a href="http://www.cs.unc.edu/Research/compgeom/gzstream/">gzstream</a> library. </p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="quick_start/" class="btn btn-neutral float-right" title="PRSice">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRSice/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
      
        <span style="margin-left: 15px"><a href="quick_start/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="search/main.js" defer></script>

</body>
</html>

<!--
MkDocs version : 1.0.4
Build Date UTC : 2020-05-20 15:48:02
-->
