<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>PRSice - PRSice-2</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "PRSice";
    var mkdocs_page_input_path = "step_by_step.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-113286405-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> PRSice-2</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Quick Start</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../quick_start/">PRSice</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../quick_start_prset/">PRSet</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Detail Guides</span></p>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">PRSice</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-data">Input Data</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#base-dataset">Base Dataset</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#target-dataset">Target Dataset</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#plink-binary">PLINK Binary</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#bgen">BGEN</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#phenotype-files">Phenotype files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#ld-reference">LD reference</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#clumping">Clumping</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#prs-calculation">PRS calculation</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#empirical-p-value-calculation">Empirical P-value calculation</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#permutation-procedure">Permutation Procedure</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#computation-algorithm">Computation Algorithm</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#output-of-results">Output of Results</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#bar-plot">Bar Plot</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#high-resolution-plot">High Resolution Plot</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#quantile-plots">Quantile Plots</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#prs-model-fit">PRS model-fit</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scores-for-each-individual">Scores for each individual</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#summary-information">Summary Information</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#log-file">Log File</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../prset_detail/">PRSet</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../command_detail/">Available Commands</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Developers</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../compilation/">Compile from Source</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../decisions/">Development Decisions</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../resources/">Useful Resources</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Misc</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../extra_steps/">Additional Steps for MAC and Window users</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../faq/">Frequently Asked Questions</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../archive/">Archive</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../update_log/">Update Log</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">PRSice-2</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Detail Guides &raquo;</li>
        
      
    
    <li>PRSice</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRSice/edit/master/docs/step_by_step.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background</h1>
<p>You will need to have basic understanding of Genome Wide Association Studies (GWAS) in order to be able to 
perform Polygenic risk score (PRS) analyses.
If you are unfamiliar with GWAS, you can consider reading <a href="https://www.ncbi.nlm.nih.gov/pubmed/29484742">this paper</a>.</p>
<h2 id="input-data">Input Data</h2>
<p>Here, we briefly discuss different input files required by PRSice:</p>
<h3 id="base-dataset">Base Dataset</h3>
<p>Base (i.e. GWAS) data must be provided as a whitespace delimited file containing association analysis results for SNPs on the base phenotype.
PRSice has no problem reading in a gzipped base file (need to have a <strong>.gz</strong> suffix).</p>
<p>If PLINK output is used, then please make sure there is a column for the effective allele (A1) and specify it with <code>--A1</code> option. </p>
<p>If your base data follows other formats, then the column headers can be provided using the <code>--chr</code>, <code>--A1</code>, <code>--A2</code>, <code>--stat</code>, <code>--snp</code>, <code>--bp</code>, <code>--pvalue</code> options</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>PRSice requires the base file to contain information of the effective allele (<code>--A1</code>), effect size estimates (<code>--stat</code>), p-value for association (<code>--pvalue</code>), and the SNP ID (<code>--snp</code>).</p>
</div>
<p>If the input file does not contain a column header, the column can be specified using their index (start counting from 0) with the <code>--index</code> flag.</p>
<p>For example, with the following input format:</p>
<table>
<thead>
<tr>
<th align="center">SNP</th>
<th align="center">CHR</th>
<th align="center">BP</th>
<th align="center">A1</th>
<th align="center">A2</th>
<th align="center">OR</th>
<th align="center">SE</th>
<th align="center">P</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">rs3094315</td>
<td align="center">1</td>
<td align="center">752566</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">0.9912</td>
<td align="center">0.0229</td>
<td align="center">0.7009</td>
</tr>
<tr>
<td align="center">rs3131972</td>
<td align="center">1</td>
<td align="center">752721</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">1.007</td>
<td align="center">0.0228</td>
<td align="center">0.769</td>
</tr>
<tr>
<td align="center">rs3131971</td>
<td align="center">1</td>
<td align="center">752894</td>
<td align="center">T</td>
<td align="center">C</td>
<td align="center">1.003</td>
<td align="center">0.0232</td>
<td align="center">0.8962</td>
</tr>
</tbody>
</table>
<p>the parameters can either be</p>
<p><code>--snp SNP --chr CHR --bp BP --A1 A1 --A2 A2 --stat OR --pvalue P</code></p>
<p>or</p>
<p><code>--snp 0 --chr 1 --bp 2 --A1 3 --A2 4 --stat 5 --pvalue 7 --index</code></p>
<p>Strand flips are automatically detected and accounted for.</p>
<p>If an imputation info score or the minor allele frequencies (MAF) are also included in the file, 
<code>--base-info &lt;Info Name&gt;:&lt;Info Threshold&gt;</code> and <code>--base-maf &lt;MAF Name&gt;:&lt;MAF Threshold&gt;</code> can be used to filter
SNPs based on their INFO score and MAF respectively.</p>
<p>For binary trait base file, SNPs can be filtered according to the MAF in case and control separately using</p>
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>--base-maf &lt;Name for Control&gt;:&lt;Threshold for Control&gt;,&lt;Name for Case&gt;:&lt;Threshold for Case&gt;
</code></pre></div>
<p>By default, PRSice will look for the following column names automatically from the base file header if <code>--index</code> was not provided or if the column name of the specific arguement(s) were not provided:</p>
<blockquote>
<p>CHR, BP, A1, A2, SNP, P, INFO (case sensitive) and OR / BETA (case insensitive)</p>
</blockquote>
<p><code>--no-default</code> can be used to disable all the defaults of PRSice.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>PRSice will ignore any columns that were not found in the base file (e.g. If<code>--A2 B</code> is specified but none of the column header is <em>B</em>,  then PRSice will treat it as if no <em>A2</em> information is presented)</p>
</div>
<h3 id="target-dataset">Target Dataset</h3>
<p>Currently two different target file format is supported by PRSice:</p>
<h4 id="plink-binary">PLINK Binary</h4>
<p>A target dataset in PLINK binary format must consist of three files: <strong>.bed</strong>, <strong>.bim</strong>, and a <strong>.fam</strong> file - where bed contains the compressed genotype data, bim contains the SNP information and fam contains the family information. Currently only SNP major PLINK format are supported  (default output of the latest PLINK program).</p>
<p>The <strong>.bed</strong> and <strong>.bim</strong> file must have the same prefix.</p>
<p>If the <strong>.fam</strong> file follow a different prefix from the <strong>.bed</strong> and <strong>bim</strong> file, it can be specified using
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>--target &lt;bim bed prefix&gt;,&lt;fam file&gt;
</code></pre></div></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The fam file MUST contains the correct number of samples or PRSice will crash</p>
</div>
<p>Missing phenotype data can be coded as NA, or -9 for binary traits and NA for quantitative traits.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>-9 will <strong>NOT</strong> be considered as missing for quantitative traits</p>
</div>
<p>If the binary file is separated into individual chromosomes, then an # can be used to specify the location of the chromosome number in the file name.
PRSice will automatically substitute # with 1-22</p>
<p>i.e. If the files are chr1.<bed|bim|fam>,chr2.<bed|bim|fam>,...,chr22.<bed|bim|fam>, just use
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>--target chr#
</code></pre></div></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Chromosome number substitution will not be performed on the external fam file as the fam file should be the same for all chromosomes. </p>
</div>
<p>Alternatively, if your PLINK files do not have a unified prefix, you can use <code>--target-list</code> to provide a file containing all
prefix to PRSice. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>.pgen</strong> files are not currently supported</p>
</div>
<h4 id="bgen">BGEN</h4>
<p>PRSice currently support BGEN v1.1 and v1.2. To specify a BGEN file, simply add the <code>--type bgen</code> or <code>--ld-type bgen</code> to the PRSice command</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In theory, we can support BGEN v1.3, but that will require us to include zstd library, developed by facebook. 
You can enable the support by including the zstd library and changing the <em>bgen_lib</em> files.</p>
</div>
<p>As BGEN does not store the phenotype information and sometime not even the sample ID, you <strong>must</strong> provide
a phenotype file (<code>--pheno</code>). Alternatively, if you have a sample file containing the phenotype information, you can 
provide it with</p>
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>--target &lt;bgen prefix&gt;,&lt;sample file&gt;
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The sample file is required even if <code>--no-regress</code> is set as the sample ID is required for output. This requirement might be losen in future versions</p>
</div>
<p>With BGEN input, a number of PRSice options become effective:</p>
<ul>
<li>
<p><code>--hard</code>: Normally, with BGEN format, PRS is calculated using the dosage information.
But hard-thresholding can be performed by using the <code>--hard</code> option. SNPs will then coded as the genotype (0,1 or 2) and filtered
according to threshold set by <code>--hard-thres</code>. If no such genotype is presented, the SNP will be
coded as missing</p>
</li>
<li>
<p><code>--hard-thres</code>: A hardcall is saved when the distance to the nearest hardcall is less than the hardcall threshold. See <a href="../command_detail/#dosage">here</a> for more detail. </p>
</li>
</ul>
<p>To perform clumping on BGEN file, we need to repeatly decompress the genotype dosage and convert them into PLINK binary format. 
To speed up the clumping process, you can allow PRSice to generate a large intermediate file, containing the hard
coded genotypes in PLINK binary format by using the <code>--allow-inter</code> option.</p>
<h3 id="phenotype-files">Phenotype files</h3>
<p>An external phenotype file can be provided to PRSice using the <code>--pheno</code> parameter. 
This must be a tab / space delimited file and missing data <strong>must</strong> be represented by either <code>NA</code> or <code>-9</code> (only for binary traits).
The first two column of the phenotype file should be the FID and the IID, or when
<code>--ignore-fid</code> is set, the first column should be the IID.
The rest of the columns can be the phenotype(s).</p>
<p>To specify a trait within the phenotype file, the column name for the trait can be specified using
<code>--pheno-col</code>, providing that the phenotype file contains a header.
Multiple column name can be provided via a comma separated
list: e.g. <code>--pheno-col A,B,C,D</code>. Trait(s) not found within the phenotype file will
be automatically skipped.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The column name(s) should not contain <em>space</em> nor <em>comma</em></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When more than one traits are provided, the column name will be appended to
the output prefix.</p>
</div>
<h2 id="ld-reference">LD reference</h2>
<p>When the target sample is small (e.g. &lt; 500 samples), an external reference panel can be used
to improve the LD estimation for clumping.</p>
<p>The LD reference follows the same notion as the target dataset. Simply use
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>--ld &lt;LD refernce&gt;
</code></pre></div>
to specify your LD reference panel file and <code>--ld-type</code> to specify the format</p>
<p>When a LD reference file is not provided and <code>--no-clump</code> is not specified, the target file
will be used as the LD reference panel</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Any parameters with the <code>--ld</code> prefix will only work on the file specified by the <code>--ld</code> parameter.
That is, if a LD reference file is not provided, none of the <code>--ld-*</code> options will be used.
If a different set of filtering is to be perforemd on the target file when performing LD calculation, it must
be provided separately to the <code>--ld</code> parameter</p>
<p>e.g. <code>--target &lt;target&gt; --ld &lt;target&gt; --keep &lt;samples for PRS&gt; --ld-keep &lt;samples for LD&gt;</code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>BGEN file will always be hard coded when used to estimate the LD</p>
</div>
<h2 id="clumping">Clumping</h2>
<p>By default, PRSice will perform Clumping to remove SNPs that are in LD with each other.
Similar to PLINK, the r<sup>2</sup> values computed by PRSice are based on maximum likelihood haplotype frequency estimates.</p>
<p>Both cases and controls are included in the LD calculation. 
Alternatively, a combination of <code>--ld</code> and <code>--ld-keep</code>/<code>-ld-remove</code> can be used to restrict LD calculation in control samples. </p>
<p>Clumping parameters can be changed by using the <code>--clump-kb</code>, <code>--clump-r2</code> and <code>--clump-p</code> option.
Clumping can be disabled using <code>--no-clump</code></p>
<h2 id="prs-calculation">PRS calculation</h2>
<p>PRSice allow different genetic models to be specified (e.g. add, dom, het, rec), and the polygenic score
of each of those are calculated differently</p>
<p>Assuming <span class="arithmatex">\(S\)</span> is the summary statistic for the effective allele and <span class="arithmatex">\(G\)</span> is the number
of the effective allele observed, then the main difference between the models is
how the genotypes are coded:</p>
<p>For additive model (add)</p>
<div class="arithmatex">\[
G = G
\]</div>
<p>For dominant model (with respect to the effective allele of the base file)</p>
<div class="arithmatex">\[
    G =
  \begin{cases}
    0 &amp; \text{if $G$ = 0} \\
    1 &amp; \text{otherwise}
  \end{cases}
\]</div>
<p>For recessive model (with respect to the effective allele of the base file)</p>
<div class="arithmatex">\[
    G =
  \begin{cases}
    1 &amp; \text{if $G$ = 2} \\
    0 &amp; \text{otherwise}
  \end{cases}
\]</div>
<p>For heterozygous model</p>
<div class="arithmatex">\[
    G =
  \begin{cases}
    1 &amp; \text{if $G$ = 1} \\
    0 &amp; \text{otherwise}
  \end{cases}
\]</div>
<p>Then depending on the <code>--score</code> option, the PRS is calculated as (assuming <span class="arithmatex">\(M_j\)</span> is the number of Alleles included in the PRS of the <span class="arithmatex">\(j^{th}\)</span> individual)</p>
<p><code>--score avg</code> (default):
$$
    PRS_j = \sum_i{\frac{S_i\times G_{ij}}{M_j}}
$$</p>
<p><code>--score sum</code>:
$$
    PRS_j = \sum_i{S_i\times G_{ij}}
$$</p>
<p><code>--score std</code>:
$$
    PRS_j = \frac{\sum_i({S_i\times G_{ij}}) - \text{Mean}(PRS)}{\text{SD}(PRS)}
$$</p>
<p><code>--score con-std</code>:
$$
    PRS_j = \frac{\sum_i({S_i\times G_{ij}}) - \text{Mean}(PRS in control)}{\text{SD}(PRS in control)}
$$
Sometimes, sample can have missing genotype. 
The <code>--missing</code> option is used to determine how
PRSice handle the missingness.</p>
<p>When not specified, the Minor Allele Frequency (MAF) in the target sample will be used as the
genotype as the sample with missing genotype.</p>
<p>If <code>--missing SET_ZERO</code> is set, the SNP for the missing
samples will be excluded.
Alternatively, if <code>--missing CENTER</code> is set, all PRS calculated will be minused by
the MAF of the SNP (therefore, missing samples will have PRS of 0).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Missingness imputation is usually based on the target samples. If you would like to 
impute the missingness using the reference sample, you can use <code>--use-ref-maf</code> parameter
to specify all MAF to be calculated using the reference samples.</p>
</div>
<h2 id="empirical-p-value-calculation">Empirical P-value calculation</h2>
<p>All approaches to PRS calculation involve parameter optimisation and are therefore overfitted. </p>
<p>There are a few methods to account for the overfitting:</p>
<ol>
<li>Evaluate performance in an independent validation sample</li>
<li>Cross validation</li>
<li>Calculate an empirical P-value</li>
</ol>
<p>In, PRSice-2, we have implemented permutation procedure to calculate the empirical P-value. </p>
<h3 id="permutation-procedure">Permutation Procedure</h3>
<p>To calculate the empirical P-value, PRSice-2 perform the following</p>
<ol>
<li>Perform standard PRSice analysis <ul>
<li>Obtain the p-value of association of the best p-value threshold (<span class="arithmatex">\(P_o\)</span>)</li>
</ul>
</li>
<li>Randomly shuffle the phenotype and repeat the PRSice analysis <ul>
<li>Obtain the p-value of association of the best p-value threshold under the null (<span class="arithmatex">\(P_{null}\)</span>)</li>
</ul>
</li>
<li>Repeat step-2 <span class="arithmatex">\(N=10,000\)</span> times (for <code>--perm 10000</code>)</li>
<li>The empirical p-value can then be calculated as</li>
</ol>
<div class="arithmatex">\[
\text{Empirical-}P = \frac{\sum_{n=1}^NI(P_{null}\lt P_o)+1}{N+1}
\]</div>
<p>where <span class="arithmatex">\(I(.)\)</span> is the indicator function. </p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>While the empirical p-value for association will be controlled for Type 1 error, 
the observed phenotypic variance explained, R<sup>2</sup>, remains unadjusted and is affected by overfitting. 
Therefore, it is imperative to perform out-of-samp,le prediction, or cross-validation to evaluate the predictive accuracy of PRS. </p>
</div>
<h3 id="computation-algorithm">Computation Algorithm</h3>
<p>In reality, PRSice-2 exploit certain property of random number generation to speed up the permutation analysis. </p>
<p>To generate random numbers, a random seed is required. When the same seed is provided, the same sequence of random number will always be generated. 
PRSice-2 exploit this property, such that the permutation analysis is performed as follow</p>
<ol>
<li>Generate the random seed or set the random seed to the user provided random seed (<span class="arithmatex">\(S\)</span>)</li>
<li>For each p-value threshold<ol>
<li>Calculate the observed p-value </li>
<li>Seed the random number generator with <span class="arithmatex">\(S\)</span></li>
<li>For Quantitative trait, (and binary trait, unless <code>--logit-perm</code> is set), decompose the matrix of the independent variables (<span class="arithmatex">\(Intercept+PRS+Covariates\)</span>)</li>
<li>Generate N copies of random phenotypes via random shuffling. </li>
<li>Calculate the p-value association for each null phenotype</li>
<li>For each permutation, check if the current null p-value is the most significant. Replace the previous "best" p-value if the current null p-value is more significant</li>
</ol>
</li>
<li>Calculate the empirical p-value once all p-value thresholds have been processed</li>
</ol>
<p>As we re-seed the random number generator for each p-value threshold, we ensure the random phenotypes generated in each p-value thresholds
are identical, allowing us to reuse the calculated PRS and the decomosed matrix, which leads to significant speed 
up of the permutation process.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>With binary traits, unless <code>--logit-perm</code> is set, we will still perform linear regression as we assume linear regression and logistic regression should produce similar t-statistics</p>
</div>
<h2 id="output-of-results">Output of Results</h2>
<h3 id="bar-plot">Bar Plot</h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Hereon,  <em>[Name]</em> is assumed to be the output prefix specified using <code>--out</code> and <em>[date]</em> is the date when the analysis was performed.</p>
</div>
<p>PRSice will always generate a bar plot displaying the model fit of the PRS at P-value threshold as indicated by <code>--bar-levels</code></p>
<p>The plot will be named as <em>[Name]_BARPLOT_[date].png</em> .
An example bar plot:</p>
<p><img alt="Bar plot" src="../img/BARPLOT.png" /></p>
<h3 id="high-resolution-plot">High Resolution Plot</h3>
<p>If <code>--fastscore</code> is not specified, a high-resolution plot named <em>[Name]_HIGH-RES_PLOT_[date].png</em> will be generated.</p>
<p>This plot present the model fit of PRS calculated at all P-value thresholds.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The model fit is defined as the <span class="arithmatex">\(R^2\)</span> of the Full model - the <span class="arithmatex">\(R^2\)</span> of the Null model</p>
<p>For example, if Sex is a covariate in the PRSice calculation, then 
model fit = <span class="arithmatex">\(R^2\)</span> of <span class="arithmatex">\(Pheno\sim PRS+Sex\)</span> - <span class="arithmatex">\(R^2\)</span> of <span class="arithmatex">\(Pheno\sim Sex\)</span></p>
</div>
<p>A green line connects points showing the model fit at the broad P-value thresholds used in the corresponding bar plot are also added.
An example high-resolution plot:</p>
<p><img alt="High Resolution Plot" src="../img/HIGH-RES_PLOT.png" /></p>
<h3 id="quantile-plots">Quantile Plots</h3>
<p>If <code>--quantile [number of quantile]</code> is specified, a quantile plot named [Name]_QUANTILE_PLOT_[date].png will be generated.
The quantile plot provide an illustration of the effect of increasing PRS on predicted risk of phenotype.
An example quantile plot:</p>
<p><img alt="Quantile Plot" src="../img/QUANTILES_PLOT.png" /></p>
<p>Specifically, the quantile plot is generated by the following steps</p>
<ol>
<li>Distribute samples into user specified number of quantiles based on their PRS</li>
<li>Treat the quantiles as a factor, where the <code>--quant-ref</code> is the base factor</li>
<li>Perform regression with <span class="arithmatex">\(Pheno \sim Quantile + Covariates\)</span> (use logistic regression if phenotype is binary, and linear regression otherwise)</li>
<li>Set the reference quantile to have coefficient of 1 (if binary) or 0 (otherwise)</li>
</ol>
<p>The point of each quantile is their OR (if binary) or coefficient (otherwise) from the regression analysis</p>
<p>A text file [Name]_QUANTILE\_[date].txt is also produced, which provides all the data used for the plotting. </p>
<p>Moreover, uneven distribution of quantiles can be specified using the <code>--quant-break</code> function, 
which will generate the strata plot. 
For example, to replicate the quantile break from Natarajan et al (2015):</p>
<table>
<thead>
<tr>
<th align="center">Percentile of PRS, %</th>
<th align="center">All studies in iCOGS excluding pKARMA <br> OR (95% CI)</th>
<th align="center">pKARMA only <br> OR (95% CI)</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">&lt;1</td>
<td align="center">0.29 (0.23 to 0.37)</td>
<td align="center">0.48 (0.28 to 0.83)</td>
</tr>
<tr>
<td align="center">&gt;1–5</td>
<td align="center">0.42 (0.37 to 0.47)</td>
<td align="center">0.48 (0.36 to 0.63)</td>
</tr>
<tr>
<td align="center">5–10</td>
<td align="center">0.55 (0.50 to 0.61)</td>
<td align="center">0.58 (0.45 to 0.74)</td>
</tr>
<tr>
<td align="center">10–20</td>
<td align="center">0.65 (0.60 to 0.70)</td>
<td align="center">0.68 (0.57 to 0.81)</td>
</tr>
<tr>
<td align="center">20–40</td>
<td align="center">0.80 (0.76 to 0.85)</td>
<td align="center">0.81 (0.71 to 0.94)</td>
</tr>
<tr>
<td align="center">40–60</td>
<td align="center">1 (referent)</td>
<td align="center">1 (referent)</td>
</tr>
<tr>
<td align="center">60–80</td>
<td align="center">1.18 (1.12 to 1.24)</td>
<td align="center">1.35 (1.19 to 1.54)</td>
</tr>
<tr>
<td align="center">80–90</td>
<td align="center">1.48 (1.39 to 1.57)</td>
<td align="center">1.56 (1.34 to 1.82)</td>
</tr>
<tr>
<td align="center">90–95</td>
<td align="center">1.69 (1.56 to 1.82)</td>
<td align="center">2.05 (1.70 to 2.47)</td>
</tr>
<tr>
<td align="center">95–99</td>
<td align="center">2.20 (2.03 to 2.38)</td>
<td align="center">2.12 (1.73 to 2.59)</td>
</tr>
<tr>
<td align="center">&gt;99</td>
<td align="center">2.81 (2.43 to 3.24)</td>
<td align="center">3.06 (2.16 to 4.34)</td>
</tr>
</tbody>
</table>
<p>The following command can be added to PRSice command:</p>
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>--quantile 100 \
--quant-break 1,5,10,20,40,60,80,90,95,99,100 \
--quant-ref 60
</code></pre></div>
<p>Specifically, <code>--quant-break</code> indicates the <strong>upper bound</strong> of each group and <code>--quant-ref</code> specify the <strong>upper bound</strong> of the reference quantiles</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The quantile boundaries are non-overlapping, with the inclusive upper bound and exclusive lower bound </p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Usually, you will need <code>--quantile 100</code> together with <code>--quant-break</code></p>
</div>
<h3 id="prs-model-fit">PRS model-fit</h3>
<p>A file containing the PRS model fit across thresholds is named <em>[Name].prsice</em>; this is stored as</p>
<blockquote>
<p>Set, Threshold, <span class="arithmatex">\(R^2\)</span>, P-value, Coefficient, Standard Deviation and Number of SNPs at this threshold</p>
</blockquote>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p><span class="arithmatex">\(R^2\)</span> reported in the prsice file is the <span class="arithmatex">\(R^2\)</span> of the Full model - the <span class="arithmatex">\(R^2\)</span> of the Null model</p>
</div>
<h3 id="scores-for-each-individual">Scores for each individual</h3>
<p>A file containing PRS for each individual at the best-fit PRS named</p>
<p><em>[Name].best</em> is provide.
This file has the format of:</p>
<blockquote>
<p>FID,IID,In_Regression, PRS at best threshold of first set, PRS at best threshold of second set, ...</p>
</blockquote>
<p>Where the In_Regression column indicate whether the sample is included in the regression model performed by PRSice.</p>
<p>If <code>--all-score</code> option is used, a file named <em>[Name].all.score</em> is also generated</p>
<p>This file has the format of </p>
<blockquote>
<p>FID, IID, PRS for first set at first threshold, PRS for first set at second threshold, ...</p>
</blockquote>
<p>If <code>--all-score</code> is used, the PRS for each individual at all threshold and all sets will be given.
In the event where the target sample size is large and a lot of threshold are tested, this file can be large.</p>
<h3 id="summary-information">Summary Information</h3>
<p>Information of the best model fit of each phenotype and gene set is stored in <em>[Name].summary</em>.
The summary file contain the following fields:</p>
<ol>
<li><strong>Phenotype</strong> - Name of Phenotype</li>
<li><strong>Set</strong> - Name of Gene Set</li>
<li><strong>Threshold</strong> - Best P-value Threshold</li>
<li><strong>PRS.R2</strong> - Variance explained by the PRS. If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Full.R2</strong> - Variance explained by the full model (including the covariates). If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Null.R2</strong> - Variance explained by the covariates. If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Prevalence</strong> - Population prevalence as indicated by the user. "-" if not provided.</li>
<li><strong>Coefficient</strong> - Regression coefficient of the model. Can provide insight of the direction of effect.</li>
<li><strong>P</strong> - P value of the model fit</li>
<li><strong>Num_SNP</strong> - Number of SNPs included in the model</li>
<li><strong>Empirical-P</strong> - Only provided if permutation is performed. This is the empirical p-value and should account for multiple testing and over-fitting</li>
</ol>
<p>Only one summary file will be generated for each PRSice run (disregarding the number of target phenotype used)</p>
<h3 id="log-file">Log File</h3>
<p>To allow for easy replication, a log file named <em>[Name].log</em> is generated for each PRSice run, 
which contain the all the commands used for the analysis and information regarding filtering, field selected etc.</p>
<p>This also allow easy identification of problems and should always be included in the bug report.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../prset_detail/" class="btn btn-neutral float-right" title="PRSet">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../quick_start_prset/" class="btn btn-neutral" title="PRSet"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRSice/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../quick_start_prset/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../prset_detail/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
