<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Shing Wan CHOI" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>PRSet - PRSice-2</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "PRSet";
        var mkdocs_page_input_path = "prset_detail.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]--> 
      <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

        ga('create', "UA-113286405-1", "auto");
        ga('send', 'pageview');
      </script>
    
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> PRSice-2
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Quick Start</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../quick_start/">PRSice</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quick_start_prset/">PRSet</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Detail Guides</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../step_by_step/">PRSice</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">PRSet</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#msigdb">MSigDB</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#gene-gtf">Gene GTF</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#bed-files">Bed Files</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#snp-set-files">SNP Set Files</a>
    </li>
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../command_detail/">Available Commands</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Developers</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../compilation/">Compile from Source</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../decisions/">Development Decisions</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../resources/">Useful Resources</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Misc</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../extra_steps/">Additional Steps for MAC and Window users</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../faq/">Frequently Asked Questions</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../archive/">Archive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../update_log/">Update Log</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">PRSice-2</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Detail Guides</li>
      <li class="breadcrumb-item active">PRSet</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/choishingwan/PRSice/edit/master/docs/prset_detail.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="input-data">Input Data</h1>
<h2 id="msigdb">MSigDB</h2>
<p>One simple way to obtain gene sets or pathway is through the <a href="http://software.broadinstitute.org/gsea/msigdb/">MSigDB</a>.
After registration in <a href="http://software.broadinstitute.org/gsea/login.jsp;jsessionid=EEFB5FCE8B9B285B2F789B46B388A647#msigdb">here</a>, you can download different gene sets curated by the Broad Institute.
Alternatively, you can also generate your own gene sets in the GMT format:</p>
<pre>
[Set A] [Gene 1] [Gene 2] ...
[Set B] [Gene 1] [Gene 2] ...
</pre>

<p>Sometimes, the MSigDB file might store the URL for the gene set on the second column</p>
<pre>
[Set A] [url for set A] [Gene 1] [Gene 2] ...
</pre>

<p>And PRSet can properly handle that.</p>
<h2 id="gene-gtf">Gene GTF</h2>
<p>As MSigDB file does not contain the genome boundary of the genes within the gene set,
one must also provide a GTF file. A GTF file contains the genome boundary of the genetic
elements within the human genome and PRSet can use the information from GTF to determine if
a SNP falls within a specific gene.
One can download the GTF file file Human (Genome build GTCh38.p7) <a href="ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens">here</a>.</p>
<p>PRSet will look for any regions with feature  of <strong>exon</strong>, <strong>gene</strong>, <strong>protein_coding</strong> or <strong>CDS</strong> (<em>case sensitive</em>).
Any genomic regions without these features will be ignored.
Alternatively, you can specify the features using the <code>--feature</code> command.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For those who are unfamiliar, different version of the genome might differ slightly in their coordinates. Therefore it is vital to ensure all the files are originated from the same genome build</p>
</div>
<h2 id="bed-files">Bed Files</h2>
<p>In addition, PRSet also accept bed file(s) as an input.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>A bed file <strong>must</strong> contain a minimum of 3 columns:</p>
</div>
<ol>
<li><strong>chrom</strong> - The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).</li>
<li><strong>chromStart</strong> - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered 0.</li>
<li><strong>chromEnd</strong> - The ending position of the feature in the chromosome or scaffold. The chromEnd base is not included in the display of the feature. For example, the first 100 bases of a chromosome are defined as chromStart=0, chromEnd=100, and span the bases numbered 0-99.</li>
</ol>
<p>PRSet will read in any number of bed files (comma separated) and use the file names as the name of the gene set.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An annoying feature of bed file is that it starts with 0 whereas for example, the plink formats starts the coordinates at 1. So do remember to -1 from the region start when you build your own bed file from scratch.</p>
</div>
<h2 id="snp-set-files">SNP Set Files</h2>
<p>Finally, PRSet also allow SNP sets, input via the <code>--snp-set</code> option.</p>
<p>Two different formats are allowed</p>
<ul>
<li>SNP list format, a file containing a single column of SNP ID. Name of the set will be the file name or can be provided using <code>--snp-set File:Name</code></li>
<li>MSigDB format: Each row represent a single SNP set with the first column containing the name of the SNP set.</li>
</ul>
<h1 id="clumping-in-prset">Clumping in PRSet</h1>
<p>In PRSice-2, clumping is performed to account for linkage disequilibrium (LD) between SNPs.
However, when performing set based analysis, special care are required to perform clumping. 
Take the following as an example:
Assume that:</p>
<ul>
<li>Light Blue fragments are the intergenic regions</li>
<li>Dark Blue fragments are the genic regions</li>
<li>Red fragments are the gene set regions</li>
<li>SNPs are represented as thunder bolt, with the "index" SNP in clumping denoted by the green thunderbolt</li>
</ul>
<p>If we simply perform a genome wide clumping, we might remove all SNPs residing within the gene set of interest, 
reducing the signal:
<img alt="Genome Wide Clumping" src="../img/genome_wide_clump.gif" /></p>
<p>Therefore, to maximize signal within each gene set, we must perform clumping for each gene sets separately:
<img alt="Set Base Clumping" src="../img/set_clump.gif" />
this can be a tedious process and are prone to error. </p>
<p>To speed up clumping, PRSice-2 adopt a "<em>capture the flag</em>" system.</p>
<p>Each SNPs contains a flag to represent their gene set membership. 
If a SNP is a member for the set, it will have a flag of 1, otherwise it will have a flag of 0.
For example: </p>
<table>
<thead>
<tr>
<th>SNP</th>
<th style="text-align: center;">Set A</th>
<th style="text-align: center;">Set B</th>
<th style="text-align: center;">Set C</th>
<th style="text-align: center;">Set D</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP 1</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td>SNP 2</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td>SNP 3</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
</tr>
</tbody>
</table>
<p>If we use SNP 1 as the index SNP, then after clumping, we will have </p>
<table>
<thead>
<tr>
<th>SNP</th>
<th style="text-align: center;">Set A</th>
<th style="text-align: center;">Set B</th>
<th style="text-align: center;">Set C</th>
<th style="text-align: center;">Set D</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP 1</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td>SNP 2</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
</tr>
<tr>
<td>SNP 3</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
</tr>
</tbody>
</table>
<p>which removes SNP 2, but will retain SNP 3. This allow us to achieve set based clumping by only performing a single pass genome wide clumping. </p>
<h1 id="p-value-threshold-and-proxy-clumping">P-value Threshold and Proxy Clumping</h1>
<h1 id="options">Options</h1>
<h2 id="proxy-prset">Proxy PRSet</h2>
<p>One complication in PRSet is the definition of SNP membership.
The default option of PRSet is to only include SNPs that are physically within the target region.
However, it is also likely for SNPs outside the region to influence functions of the set.
Therefore we provide the <code>--proxy</code> option. Essentially, this provide a soft cutoff to SNP membership.
For example, when user define <code>--proxy 0.8</code>, if LD between SNP A and SNP B is more than 0.8, then SNP A will be considered to be within the same regions as SNP B and vice versa.</p>
<h2 id="p-value-thresholding">P-value thresholding</h2>
<p>By default, PRSet do not perform p-value thresholding and will simply calculate the set based PRS at P-value threshold of 1. 
This is because it is unclear whether the set is associated with the phenotype when the best-threshold contained only a small portion of SNPs within the gene sets.
If you wish to perform p-value thresholding with PRSet, you will need to specify any of the parameters related to p-value thresholding, i.e. <code>--interval</code>, <code>--lower</code>, <code>--upper</code>, <code>--fastscore</code> or <code>--bar-levels</code>.</p>
<h1 id="competitive-p-value-calculation">Competitive P-value Calculation</h1>
<p>A challenge in Set base analysis is to obtain a competitive p-value, 
which indicates the level of enrichment, as opposed to the self-contained 
p-value which indicates the level of association. 
To obtain a competitive p-value, PRSet can perform a permutation analysis as follow</p>
<ol>
<li>Allocate SNPs to each gene sets</li>
<li>Allocate SNPs to a background gene set<ul>
<li>if <code>--full-back</code> is specified, use the whole genome as the background</li>
<li>if a background file is provided via the <code>--background</code> command, it will be used to construct the background set</li>
<li>otherwise, will try to use the GTF file provided from <code>--gtf</code> command as the background (with feature filtering w.r.t <code>--feature</code>)</li>
</ul>
</li>
<li>Perform <a href="./#clumping-in-prset">set based clumping</a> on all sets (including the background set)</li>
<li>Obtain the p-value of association for the best threshold for each sets (<span class="arithmatex">\(P_{observed}\)</span>)<ul>
<li>While PRSet allow one to perform p-value thresholding on the set scores, we recommend against it as it is difficult to interpret the result. Using an extreme example, if only one SNP is included in the best threshold for a set, should we really consider this single SNP as representative of the gene set?</li>
</ul>
</li>
<li>For each gene set with <span class="arithmatex">\(N\)</span> post-clump SNPs<ol>
<li>Randomly select <span class="arithmatex">\(N\)</span> post-clump SNPs from the background set and construct a null PRS</li>
<li>Calculate the p-value of association of the null PRS to obtain a null P-value (<span class="arithmatex">\(P_{null}\)</span>)</li>
<li>Repeat 1-2 <span class="arithmatex">\(M\)</span> times, where <span class="arithmatex">\(M\)</span> can be set via <code>--set-perm</code></li>
<li>The competitive P-value is calculated as
    $$
\text{Competitive-}P = \frac{\sum_{n=1}^NI(P_{null}\lt P_observed)+1}{N+1}
    $$
    where <span class="arithmatex">\(I(.)\)</span> is the indicator function. </li>
</ol>
</li>
</ol>
<h2 id="computation-algorithm">Computation Algorithm</h2>
<p>Due to the number of operation required, the set based permutation are extremely time consuming. To speed up the set based permutation, we noted that in regression,</p>
<p>$$
Y\sim X\beta+C+\epsilon
$$
and </p>
<div class="arithmatex">\[
X\sim Y\beta+C+\epsilon
\]</div>
<p>will generate the same t-statistic for <span class="arithmatex">\(X\)</span> in the first equation and <span class="arithmatex">\(Y\)</span> in the second equation. Based on this observation, we can then do the following</p>
<ol>
<li>Generate a matrix <span class="arithmatex">\(A\)</span> containing the phenotype of interest and the covariates</li>
<li>Decompose matrix <span class="arithmatex">\(A\)</span></li>
<li>For each new PRS calculated, solve <span class="arithmatex">\(PRS=A\beta+\epsilon\)</span> and obtain the t-statistic. These t-statistics are then used to construct the null distribution, allow us to obtain the competitive p-value</li>
</ol>
<p>As we only need to do the decomposition once, this should significantly increase the speed of set based permutation. In our test, for the TOY data, with <code>--set-perm 5000</code>, we can speed up the set-based permutation by around 20~25%</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>With binary traits, unless <code>--logit-perm</code> is set, we will still perform linear regression as we assume linear regression and logistic regression should produce similar t-statistics</p>
</div>
<h1 id="output-data">Output Data</h1>
<h2 id="prs-model-fit">PRS model-fit</h2>
<p>A file containing the PRS model fit across thresholds is named <strong>[Name].prsice</strong>, where <strong>[Name]</strong> is the
output prefix name as specified by <code>--out</code> this is stored as</p>
<p>Name of Set, Threshold, R2, P-value, Coefficient, Standard Error, and Number of SNPs at this threshold</p>
<h2 id="scores-for-each-individual">Scores for each individual</h2>
<p>A file containing PRS for each individual at the best-fit PRS named</p>
<p><strong>[Name].best</strong> is provide.
This file has the format of:</p>
<p>FID,IID, In Regression, PRS at best threshold for Set 1, PRS at best threshold for Set 2, ...</p>
<p>Where the has phenotype column indicate whether the sample contain all
the required phenotype for PRSice analysis (e.g. Samples with missing
phenotype/covariate will not be included in the regression.
These samples will be indicated as "No" under the in regression column)</p>
<p>If <code>--all</code> option is used, a file named</p>
<p><strong>[Name].all.score</strong> is also generated</p>
<p>Please note, if <code>--all</code> options is used, the PRS for each individual at all threshold will be given.
In the event where the target sample size is large and a lot of threshold are tested, this file can be large.
This is especially true when large number of gene sets were provided.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>PRSice also supports multiple phenotypes for target data.
All output prefix will change to [Name].[Pheno] where [Pheno]
is the name of the phenotype. 
For more details on the options used to implement this, see
<a href="../step_by_step/#phenotype-files">here</a>.</p>
</div>
<h2 id="summary-information">Summary Information</h2>
<p>Information of the best model fit of each phenotype and gene set is stored in [Name].summary.
The summary file contain the following fields:</p>
<ol>
<li><strong>Phenotype</strong> - Name of Phenotype</li>
<li><strong>Set</strong> - Name of Gene Set</li>
<li><strong>Threshold</strong> - Best P-value Threshold</li>
<li><strong>PRS.R2</strong> - Variance explained by the PRS. If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Full.R2</strong> - Variance explained by the full model (including the covariates). If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Null.R2</strong> - Variance explained by the covariates. If prevalence is provided, this will be adjusted for ascertainment</li>
<li><strong>Prevalence</strong> - Population prevalence as indicated by the user. "-" if not provided.</li>
<li><strong>Coefficient</strong> - Regression coefficient of the model. Can provide insight of the direction of effect.</li>
<li><strong>P</strong> - P value of the model fit</li>
<li><strong>Num_SNP</strong> - Number of SNPs included in the model</li>
<li><strong>Empirical-P</strong> - Only provided if permutation is performed. This is the empirical p-value and should account for multiple testing and over-fitting</li>
<li><strong>Competitive-P</strong> - Only provided if set permutation is performed. This is the competitive p-value and should measure the enrichment of signal of the gene set</li>
</ol>
<h2 id="multi-set-plot">Multi-Set Plot</h2>
<p>When the <code>--multi-plot &lt;N&gt;</code> option is set, the results of the top <em>N</em> gene sets will be plotted.
An example of the multi-set plot is:</p>
<p><img alt="multo set plot" src="../img/MULTISET_BARPLOT.png" /></p>
<h2 id="other-figures">Other Figures</h2>
<p>The default behaviour of PRSet is to only plot the High-resolution plot, bar-plot and the quantile plot
for the "Base" data. You can change this behaviour by using the <code>--plot-set</code> option.</p>
<h2 id="log-file">Log File</h2>
<p>We value reproducible research. Therefore we try our best to make replicating PRSice run easier.
For every PRSice run, a log file named <strong>[Name].log</strong> is generated which contain the all the commands
used for the analysis and information regarding filtering, field selected etc.</p>
<p>This also allow users to quickly identify problems in the input dataset.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../step_by_step/" class="btn btn-neutral float-left" title="PRSice"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../command_detail/" class="btn btn-neutral float-right" title="Available Commands">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/choishingwan/PRSice" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../step_by_step/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../command_detail/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../javascripts/details.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
